daffodil-tdml-processor/src/test/scala/org/apache/daffodil/processor/tdml/TestTDMLRunner2.scala

```diff
@@ -677,4 +676,80 @@ abc # a comment
     runner.reset
   }
 
+  // DFDL-2947
+  @Test def test_rootNSSpecifiesNoNamespaceRoot(): Unit = {
+    val tdmlTestSuite =
+      <tdml:testSuite suiteName="theSuiteName" xmlns:ex={example} xmlns:tdml={tdml} xmlns:dfdl={
+        dfdl
+      } xmlns:xsd={xsd} xmlns:xs={xsd} xmlns:xsi={xsi}>
+        <tdml:parserTestCase name="test1" root="data" rootNS="" model="/test/tdml/chameleon-schema1.dfdl.xsd">
+          <tdml:document>37</tdml:document>
+          <tdml:infoset>
+            <tdml:dfdlInfoset>
+              <data>37</data>
+            </tdml:dfdlInfoset>
+          </tdml:infoset>
+        </tdml:parserTestCase>
+      </tdml:testSuite>
+
+    val runner = new Runner(tdmlTestSuite)
+    runner.runOneTest("test1")
+  }
+
+  @Test def test_rootNSSpecifiesNamespaceRoot(): Unit = {
+    val tdmlTestSuite =
+      <tdml:testSuite suiteName="theSuiteName" xmlns:ex={example} xmlns:tdml={tdml} xmlns:dfdl={
+        dfdl
+      } xmlns:xsd={xsd} xmlns:xs={xsd} xmlns:xsi={xsi}>
+        <tdml:parserTestCase name="test1" root="data" rootNS="urn:schema:namespace" model="/test/tdml/chameleon-schema1.dfdl.xsd">
+          <tdml:document>boy</tdml:document>
+          <tdml:infoset>
+            <tdml:dfdlInfoset>
+              <data>boy</data>
+            </tdml:dfdlInfoset>
+          </tdml:infoset>
+        </tdml:parserTestCase>
+      </tdml:testSuite>
+
+    val runner = new Runner(tdmlTestSuite)
+    runner.runOneTest("test1")
+  }
+
+  @Test def test_noRootCandidates1(): Unit = {
+    val tdmlTestSuite =
+      <tdml:testSuite suiteName="theSuiteName" xmlns:ex={example} xmlns:tdml={tdml} xmlns:dfdl={
+        dfdl
+      } xmlns:xsd={xsd} xmlns:xs={xsd} xmlns:xsi={xsi}>
+        <tdml:parserTestCase name="test1" root="data1" model="/test/tdml/chameleon-schema1.dfdl.xsd">
+          <tdml:document>37</tdml:document>
+          <tdml:errors>
+          <tdml:error>Schema Definition Error</tdml:error>
+          <tdml:error>no root element found</tdml:error>
+          <tdml:error>data1</tdml:error>
+          </tdml:errors>
+        </tdml:parserTestCase>
+      </tdml:testSuite>
+
+    val runner = new Runner(tdmlTestSuite)
+    runner.runOneTest("test1")
+  }
+
+  @Test def test_noRootCandidates2(): Unit = {
+    val tdmlTestSuite =
+      <tdml:testSuite suiteName="theSuiteName" xmlns:ex={example} xmlns:tdml={tdml} xmlns:dfdl={
+        dfdl
+      } xmlns:xsd={xsd} xmlns:xs={xsd} xmlns:xsi={xsi}>
+        <tdml:parserTestCase name="test1" root="data" rootNS="doesNotExist" model="/test/tdml/chameleon-schema1.dfdl.xsd">
+          <tdml:document>37</tdml:document>
+          <tdml:errors>
+          <tdml:error>Schema Definition Error</tdml:error>
+          <tdml:error>no global element found</tdml:error>
+          <tdml:error>{"{doesNotExist}"}data</tdml:error>
```

Comment:
This error message looks kind of werid, seems there is an extra set of curly braces and quotes? I would think the error should just be something like `<tdml:error>{doesNotExist}data</tdml:error>`

Reply:
Oh that was done that way because we are in a scala file, and the braces in scalaxml mean interpolation, so it was done the way it was to sort of escape the braces, but I'll update to the official way to escape which is {{ and }}